Add DeviceRegistryRestServerTest

calohmn opened the pull request:
Test for `DeviceRegistryRestServer`.

pellmann commented:
I think @sysexcontrol just change the JSON-names in the payload from device_id to device-id, to match with all other usage in JSONs. I think we use device_id only in for AMQP application properties. Anyway it may be better to use on of the already defined constants for this?

calohmn commented:
Good point, I've extracted the "device_id" request param to a constant in the test (amended the commit).
The already defined constant for this would be `RegistrationHttpEndpoint.PARAM_DEVICE_ID`. That field is private though.

Regarding the inconsitency concerning usage of "device_id" in the request JSON and "device-id" in the response JSON, I've created #326.
As `RegistrationHttpEndpoint.PARAM_DEVICE_ID` probably becomes obsolete with the fix for #326, I think we should wait for that fix and then use the corresponding constant in the `DeviceRegistryRestServer`.

calohmn commented:
Now that #326 has been fixed and "device_id" was changed to "device-id", I've adapted the constant being used in the test (rebased and sqashed).

calohmn commented:
I've rebased the branch and applied recent changes to return codes in the REST API. 
As FormData support will most probably be dropped (see #316), I've removed the corresponding test methods.

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+
+    @BeforeClass
+    public static void setUp(final TestContext context) throws Exception {
+        vertx = Vertx.vertx();
```
throws Exception : IMHO can be deleted?
Or is there a good reason for declaring them at all methods in this file?

calohmn replied:
It wasn't added on purpose, the IDEA template for adding a test method included it. 
I'll remove them.

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testAddDeviceWithoutDeviceId(final TestContext context) throws Exception {
```
delete throw statement?

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testAddDisabledDevice(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testAddDeviceFailsAlreadyExists(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testAddDeviceFailsMissingContentType(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testAddDeviceFailsMissingBody(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testGetDevice(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testUpdateDevice(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testUpdateDeviceWithDisablingDevice(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testUpdateDeviceFailsNoSuchDevice(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testUpdateDeviceWithEmptyRequestBody(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testUpdateDeviceFailsMissingContentType(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testRemoveDevice(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Test
+    public void testRemoveDeviceFailsNoSuchDevice(final TestContext context) throws Exception {
```
again

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+    @Test
+    public void testAddDevice(final TestContext context) throws Exception {
+        final String requestUri = String.format("/%s/%s", REGISTRATION_ENDPOINT, TENANT);
+        final String requestBody = "{\"" + FIELD_DEVICE_ID + "\":\"" + DEVICE_ID + "\",\"test\":\"test\"}";
```
A general proposal:
to improve readability I would rather use String.format.

E.g. like the following:
`
final String requestBody2 = String.format("{\"%s\":\"%s\",\"test\":\"test\"}", FIELD_DEVICE_ID, DEVICE_ID);
`

or (even more readable):
`
final String requestBody3 = String.format("{'%s':'%s','test':'test'}", FIELD_DEVICE_ID, DEVICE_ID).replace('\'','"');
`

WDYT?

calohmn replied:
Yes, that string concatenating/quoting better be improved.
I like your 2nd solution. In this scenario of creating a JSON string, there's also the possibility to use `JsonObject`:
`new JsonObject().put(FIELD_DEVICE_ID, DEVICE_ID).put("test", "test").encode();`.
I think that's even more readable.

sysexcontrol replied:
:+1: That is indeed even better!

sophokles73 replied:
You should use `JsonObject.encodePrettily()`, though. Otherwise the whole JSON be on one line ...

calohmn replied:
But I think that isn't an issue here. For the cases here, the JSON is short, containing only 2-3 key-value pairs, so during debugging the string value would still be readable.

sysexcontrol commented on services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java:
```diff
@@ -0,0 +1,430 @@
+        final Async async = context.async();
+        vertx.createHttpClient().post(getPort(), HOST, requestUri).putHeader("content-type", CONTENT_TYPE_JSON)
+                .handler(response -> {
+                    context.assertEquals(201, response.statusCode());
```
Maybe better use `HttpURLConnection.HTTP_CREATED`?
Same for all occurences of direct status code below....

sysexcontrol commented:
One more thing : I think the class should better be renamed to RegistrationRestServerTest.
We will soon have the same for the (then REST based) Credentials API, which resides in the device registry component as well, so the current naming is not well matching anymore.

calohmn commented:
I've done the proposed changes (rebased and squashed).
